### PR TITLE
Add a macro for punning keyword arguments

### LINKED ISSUE #1119: [feature] Shorthand for `:kwarg True`, `:kwarg False`, and `:kwarg kwarg`
It's common to want to set a keyword argument to a Boolean, and also common to want to set a keyword argument to the value of a variable with the same name. I propose the following syntactic sugar for keywords when they're used for keyword arguments:
- Translate `:+foo` to `:foo True`
- Translate `:!foo` to `:foo False`
- Translate `:^foo` to `:foo foo`

(One might also be tempted to add a construct for `:foo None`, but this seems like an unusual enough case that having Hy programmers learn another piece of shorthand isn't worth it.)

I included the first two translations (`:+foo`) and (`:!foo`) in [my own implementation of keyword arguments for Hy](https://github.com/Kodiologist/Kodhy/blob/88631400ce/kodhy/macros.hy#L32), which I made because Hy didn't have that feature built-in yet. I've found this sugar handy and use it frequently in practice. It's a bit like `--flag` and `--no-flag` in command-line arguments.

 I haven't previously used the last translation (`:^foo`). It's similar to a GHC extension to Haskell called ["record puns"](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#record-puns).

### PR DESCRIPTION
**Pull request by Kodiologist**
Closes #1119.

Replaces  #1685.

### CODE REVIEW

**Comment by gilch** [comment]
Putting this in core seems premature. It's an experimental macro not even based on anything from Clojure or Common Lisp, that means it belongs in contrib. I don't imagine I'd use this enough to justify it, but if it proves popular we could then move it to extra.

File: hy/core/macros.hy [hunk 1]
```diff
@@ -276,3 +276,10 @@ Such 'o!' params are available within `body` as the equivalent 'g!' symbol."
                 _hy_tag
                 [None]
                 ['~symbol])))))
+
+
+(defmacro pun [&rest body]
+  (HyExpression (reduce + (gfor x body
+    (if (and (keyword? x) (.startswith x.name "^"))
+      [(HyKeyword (cut x.name 1)) (HySymbol (cut x.name 1))]
+      [x])))))
```
Comment (gilch) [inline hy/core/macros.hy]: This definition isn't indented properly. Make sure it passes parlinter.

Comment (Kodiologist) [reply hy/core/macros.hy]: [I asked you to stop this.](https://github.com/hylang/hy/pull/1338#discussion_r131307803) [Twice.](https://github.com/hylang/hy/pull/1378#discussion_r132749045) [Then you insulted me about it.](https://github.com/hylang/hy/pull/1614#issuecomment-391211451) Please don't comment on my indentation again.

**Comment by Kodiologist** [comment]
I wouldn't say it's experimental (it's a very simple macro), and the inclusion criteria for contrib and extra have nothing to do with Common Lisp or Clojure.

**Comment by brandonwillard** [comment]
What's the inclusion criteria for `hy.core.macros`?

**Comment by Kodiologist** [comment]
We never decided on any. In practice, it contains macros that are short and are of general use, and we should probably stick to that.

**Comment by Kodiologist** [comment]
@gilch To be clear, are you vetoing this as-is?

**Comment by Kodiologist** [comment]
It's clear that lots of time and people and thinking went into previous languages, but none of those things imply that the results are any good. People's decisions are subject to inertia and groupthink, and there's no group so large that anything important is guaranteed to be thought of by somebody in the group. How things have been done previously is not any indiciation as to what should be done.

> Are we sure we want a macro rather than a tag?

The distinction doesn't seem important. I could switch it to a tag macro if you'd prefer that.

> How do we work around it?

With mangling, as you just realized and as already tested. It's clumsy, but it's hard to see how the clumsiness will matter considering the sheer weirdness of beginning a function parameter's name with a caret, and the coincidence of wanting to use `pun` in the same call.

> Maybe `(pun spam :^foo :^bar :XhatXq q)` oh, but you've vetoed those mangle improvements #1651

To be clear, I'm not against mangling nicknames for ASCII characters, just the absence of a clear marker of whether a name has been mangled (which is currently the `hyx_` prefix).

> I don't know how well this would work in practice. And, emphatically, *neither do you*.

I'd anticipated this problem you thought of, but if you can think of any other potential problems with the macro, by all means mention them and I can fix them and add tests. It's appropriate to hold up a PR for known problems, but not for problems they might have but we haven't even thought of yet.

**Comment by Kodiologist** [comment]
@gilch Have you thought of any potential problems?

**Comment by Kodiologist** [comment]
@gilch So what's the word? Do you want it as a tag macro instead?

**Comment by alphapapa** [comment]
Although I've known of Hy for a while, I haven't used it "in anger" much, so I'm still a tyro.  My perspective comes from using other Lisps, so for whatever it's worth:

Since `apply` was removed (which I miss, but can adapt to), the proposed syntax seems strange to me, since it would work roughly like `apply`:

```hy
(apply dict [:monitor monitor :verbose verbose :cv-folds cv-folds])
(pun dict :^monitor :^verbose :^cv-folds)
```

What if `pun` worked like the replacement for `apply`, i.e. using `#**`, something like:

```hy
(dict #** (pun monitor verbose cv-folds))
```

**Pull request closed without merging**